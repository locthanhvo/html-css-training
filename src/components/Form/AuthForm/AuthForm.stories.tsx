import { StoryObj, Meta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import AuthForm from '.';

const meta: Meta<typeof AuthForm> = {
  title: 'Components/AuthForm',
  tags: ['autodocs'],
  component: AuthForm,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof AuthForm>;

export const SignIn: Story = {
  args: {},
};

export const SignUp: Story = {
  args: {
    isRegister: true,
  },
};
