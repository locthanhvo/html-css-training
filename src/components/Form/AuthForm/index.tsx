import { ChangeEvent, memo, useCallback } from 'react';
import { Checkbox, Flex, Text } from '@chakra-ui/react';
import { Controller, useForm } from 'react-hook-form';
import { Link } from 'react-router-dom';

// Components
import { ForwardIcon } from '@/components/Icons';
import { InputField, ActionButton } from '@/components';

// Constants
import { AUTH_SCHEMA, PUBLIC_ROUTERS } from '@/constants';

// Types
import { TAuthForm } from '@/types';
import { displayFieldAuthForm } from '@/utils';

interface AuthFormProps {
  isRegister?: boolean;
  isLoading?: boolean;
  onSubmit: (data: TAuthForm) => void;
}

const AuthForm = ({
  isLoading = false,
  isRegister = false,
  onSubmit,
}: AuthFormProps) => {
  const {
    control,
    formState: { isSubmitting, isValid },
    handleSubmit,
    clearErrors,
  } = useForm<TAuthForm>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    defaultValues: {
      email: '',
      password: '',
      ...(isRegister && {
        phone: '',
        confirmPassword: '',
        isAcceptPrivacyPolicy: false,
      }),
    },
  });

  const handleInputChange = useCallback(
    ({
      field,
      data,
      isError,
      onChange,
    }: {
      field: keyof TAuthForm;
      data: string;
      isError: boolean;
      onChange: (value: string) => void;
    }) => {
      isError && clearErrors(field);
      onChange(data);
    },
    [clearErrors],
  );

  return (
    <Flex
      as="form"
      w="full"
      direction="column"
      gap={4}
      onSubmit={handleSubmit(onSubmit)}
    >
      {displayFieldAuthForm(isRegister).map(
        ({ name, rules, type, label, placeholder, ariaLabel }) => (
          <Controller
            key={name}
            control={control}
            rules={rules}
            name={name}
            render={({ field, fieldState: { error } }) => (
              <InputField
                type={type}
                label={label}
                placeholder={placeholder}
                {...field}
                isError={!!error}
                errorMessages={error?.message}
                isDisabled={isSubmitting}
                onChange={(data) =>
                  handleInputChange({
                    field: name,
                    data,
                    isError: !!error,
                    onChange: field.onChange,
                  })
                }
                aria-label={ariaLabel}
              />
            )}
          />
        ),
      )}

      {isRegister ? (
        <Controller
          control={control}
          rules={AUTH_SCHEMA.AGREE_POLICY}
          name="isAcceptPrivacyPolicy"
          render={({ field: { value, onChange } }) => (
            <Checkbox
              isChecked={value}
              onChange={(e: ChangeEvent<HTMLInputElement>) =>
                onChange(e.target.checked)
              }
            >
              <Text fontSize="base" fontWeight="regular">
                I accept{' '}
                <Text as="span" color="blue.300">
                  the terms and conditions of use.
                </Text>
              </Text>
            </Checkbox>
          )}
        />
      ) : (
        <Text textAlign="center" fontSize="base" fontWeight="regular">
          Forgot your password ?
        </Text>
      )}

      <ActionButton
        type="submit"
        size="xl"
        title={isRegister ? 'Sign Up' : 'Sign In'}
        bgColor="blue.200"
        rightIcon={!isRegister ? <ForwardIcon /> : undefined}
        isDisabled={!isValid && !isSubmitting}
        isLoading={isLoading || isSubmitting}
      />

      <Text textAlign="center" fontSize="base" fontWeight="regular">
        {isRegister ? (
          <>
            Already have an account?
            <Link to={PUBLIC_ROUTERS.SIGN_IN} color="blue.300">
              <Text as="span" color="blue.300">
                {' '}
                To log in
              </Text>
            </Link>
          </>
        ) : (
          <>
            You do not have an account ?
            <Link to={PUBLIC_ROUTERS.SIGN_UP}>
              <Text as="span" color="blue.300">
                {' '}
                Create an account
              </Text>
            </Link>
          </>
        )}
      </Text>
    </Flex>
  );
};

export default memo(AuthForm);
