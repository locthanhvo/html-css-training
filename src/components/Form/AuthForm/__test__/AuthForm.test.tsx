import { render, screen } from '@testing-library/react';

// Component
import AuthForm from '..';
import { BrowserRouter } from 'react-router-dom';

const mockOnSubmit = jest.fn();

describe('AuthForm components', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('match snapshot with login form', () => {
    const { container } = render(
      <BrowserRouter>
        <AuthForm onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('match snapshot with register form', () => {
    const { container } = render(
      <BrowserRouter>
        <AuthForm isRegister onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it('renders Login form by default', () => {
    render(
      <BrowserRouter>
        <AuthForm onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(screen.getByRole('button', { name: 'Sign In' })).toBeInTheDocument();
  });

  it('renders Register form by default', () => {
    render(
      <BrowserRouter>
        <AuthForm isRegister onSubmit={mockOnSubmit} />
      </BrowserRouter>,
    );

    expect(screen.getByRole('button', { name: 'Sign Up' })).toBeInTheDocument();
  });
});
