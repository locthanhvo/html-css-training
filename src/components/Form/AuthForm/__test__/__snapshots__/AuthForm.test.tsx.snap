// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthForm components match snapshot with login form 1`] = `
<div>
  <form
    class="css-jbe4xg"
  >
    <div
      aria-label="email"
      class="chakra-form-control css-0"
      name="email"
      placeholder="Email"
      role="group"
      value=""
    >
      <label
        class="chakra-form__label css-zecyy3"
        for="field-:r0:"
        id="field-:r0:-label"
      >
        Email
      </label>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          aria-label="email"
          class="chakra-input css-adz8kx"
          id="field-:r0:"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      aria-label="password"
      class="chakra-form-control css-0"
      name="password"
      placeholder="Password"
      role="group"
      value=""
    >
      <label
        class="chakra-form__label css-zecyy3"
        for="field-:r1:"
        id="field-:r1:-label"
      >
        Password
      </label>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          aria-label="password"
          class="chakra-input css-adz8kx"
          id="field-:r1:"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <p
      class="chakra-text css-1tr1rgu"
    >
      Forgot your password ?
    </p>
    <button
      class="chakra-button css-ey593w"
      disabled=""
      name="Sign In"
      type="submit"
    >
      Sign In
      <span
        class="chakra-button__icon css-1hzyiq5"
      >
        <svg
          fill="none"
          height="12"
          viewBox="0 0 20 12"
          width="20"
        >
          <path
            d="M19.2688 4.00107L16.0434 0.742375C15.966 0.66426 15.8738 0.602257 15.7722 0.559946C15.6707 0.517634 15.5617 0.49585 15.4517 0.49585C15.3417 0.49585 15.2327 0.517634 15.1312 0.559946C15.0296 0.602257 14.9374 0.66426 14.86 0.742375C14.7047 0.898528 14.6176 1.10976 14.6176 1.32994C14.6176 1.55012 14.7047 1.76135 14.86 1.9175L17.827 4.9095H0.833425C0.612387 4.9095 0.400402 4.99731 0.244105 5.1536C0.087807 5.3099 0 5.52189 0 5.74293H0C0 5.96396 0.087807 6.17595 0.244105 6.33225C0.400402 6.48854 0.612387 6.57635 0.833425 6.57635H17.877L14.86 9.58502C14.7819 9.66249 14.7198 9.75467 14.6775 9.85623C14.6352 9.95779 14.6134 10.0667 14.6134 10.1767C14.6134 10.2868 14.6352 10.3957 14.6775 10.4973C14.7198 10.5988 14.7819 10.691 14.86 10.7685C14.9374 10.8466 15.0296 10.9086 15.1312 10.9509C15.2327 10.9932 15.3417 11.015 15.4517 11.015C15.5617 11.015 15.6707 10.9932 15.7722 10.9509C15.8738 10.9086 15.966 10.8466 16.0434 10.7685L19.2688 7.53479C19.737 7.06599 20 6.4305 20 5.76793C20 5.10535 19.737 4.46987 19.2688 4.00107Z"
            fill="white"
          />
        </svg>
      </span>
    </button>
    <p
      class="chakra-text css-1tr1rgu"
    >
      You do not have an account ?
      <a
        href="/sign-up"
      >
        <span
          class="chakra-text css-z2nxer"
        >
           
          Create an account
        </span>
      </a>
    </p>
  </form>
</div>
`;

exports[`AuthForm components match snapshot with register form 1`] = `
<div>
  <form
    class="css-jbe4xg"
  >
    <div
      aria-label="email"
      class="chakra-form-control css-0"
      name="email"
      placeholder="Email"
      role="group"
      value=""
    >
      <label
        class="chakra-form__label css-zecyy3"
        for="field-:r2:"
        id="field-:r2:-label"
      >
        Email
      </label>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          aria-label="email"
          class="chakra-input css-adz8kx"
          id="field-:r2:"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      aria-label="phone"
      class="chakra-form-control css-0"
      name="phone"
      placeholder="Phone"
      role="group"
      value=""
    >
      <label
        class="chakra-form__label css-zecyy3"
        for="field-:r3:"
        id="field-:r3:-label"
      >
        Phone
      </label>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          aria-label="phone"
          class="chakra-input css-adz8kx"
          id="field-:r3:"
          name="phone"
          placeholder="Phone"
          type="number"
          value=""
        />
      </div>
    </div>
    <div
      aria-label="password"
      class="chakra-form-control css-0"
      name="password"
      placeholder="Password"
      role="group"
      value=""
    >
      <label
        class="chakra-form__label css-zecyy3"
        for="field-:r4:"
        id="field-:r4:-label"
      >
        Password
      </label>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          aria-label="password"
          class="chakra-input css-adz8kx"
          id="field-:r4:"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      aria-label="confirmPassword"
      class="chakra-form-control css-0"
      name="confirmPassword"
      placeholder="Confirm password"
      role="group"
      value=""
    >
      <label
        class="chakra-form__label css-zecyy3"
        for="field-:r5:"
        id="field-:r5:-label"
      >
        Confirm password
      </label>
      <div
        class="chakra-input__group css-bx0blc"
        data-group="true"
      >
        <input
          aria-label="confirmPassword"
          class="chakra-input css-adz8kx"
          id="field-:r5:"
          name="confirmPassword"
          placeholder="Confirm password"
          type="password"
          value=""
        />
      </div>
    </div>
    <label
      class="chakra-checkbox css-1elxtiq"
    >
      <input
        class="chakra-checkbox__input"
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
        type="checkbox"
        value=""
      />
      <span
        aria-hidden="true"
        class="chakra-checkbox__control css-dnty2r"
      />
      <span
        class="chakra-checkbox__label css-1oeb2oe"
      >
        <p
          class="chakra-text css-12t4bqy"
        >
          I accept
           
          <span
            class="chakra-text css-z2nxer"
          >
            the terms and conditions of use.
          </span>
        </p>
      </span>
    </label>
    <button
      class="chakra-button css-ey593w"
      disabled=""
      name="Sign Up"
      type="submit"
    >
      Sign Up
    </button>
    <p
      class="chakra-text css-1tr1rgu"
    >
      Already have an account?
      <a
        color="blue.300"
        href="/sign-in"
      >
        <span
          class="chakra-text css-z2nxer"
        >
           
          To log in
        </span>
      </a>
    </p>
  </form>
</div>
`;
